src/core/cgroup.c

```diff
@@ -556,7 +558,40 @@ static uint64_t cgroup_context_cpu_shares(CGroupContext *c, ManagerState state)
                 return CGROUP_CPU_SHARES_DEFAULT;
 }
 
-static void cgroup_apply_unified_cpu_config(Unit *u, uint64_t weight, uint64_t quota) {
+uint64_t cgroup_cpu_adjust_period(uint64_t period, uint64_t quota, uint64_t resolution, uint64_t max_period) {
+        /* kernel uses a minimum resolution of 1ms, so both period and (quota * period)
+         * need to be higher than that boundary. quota is specified in USecPerSec.
+         * Additionally, period must be at most max_period. */
+        assert(quota > 0);
+
+        return MIN(MAX3(period, resolution, resolution * USEC_PER_SEC / quota), max_period);
+}
+
+static uint64_t cgroup_cpu_adjust_period_and_log(Unit *u, uint64_t period, uint64_t quota) {
+        uint64_t new_period;
+
+        if (quota == USEC_INFINITY)
+                /* Always use default period for infinity quota. */
+                return CGROUP_CPU_QUOTA_DEFAULT_PERIOD_USEC;
+
+        if (period == USEC_INFINITY)
+                /* Default period was requested. */
+                period = CGROUP_CPU_QUOTA_DEFAULT_PERIOD_USEC;
+
+        /* Clamp to interval [1ms, 1s] */
+        new_period = cgroup_cpu_adjust_period(period, quota, USEC_PER_MSEC, USEC_PER_SEC);
+
+        if (new_period != period) {
+                char v[FORMAT_TIMESPAN_MAX];
+                log_unit_full(u, LOG_WARNING, 0,
+                              "Clamping CPU interval for cpu.max: period is now %s",
+                              format_timespan(v, sizeof(v), new_period, 1));
+        }
```

Comment:
hmm, so this way we'll log at LOG_WARNING whenever we realize the cgroup, which might be quite often. We really should generate this message only once though I think (and downgrade to LOG_DEBUG from then on). Maybe add a bool flag to the unit stores whether we already warned, and if so, doesn't warn again?

Reply:
This one hasn't been addressed yet. The problem with "already warned" is how to clear it every time the value is updated... I've been thinking of storing the last clamped value instead, that way I can compare to it and skip warning if it still matches. Will update with a separate commit on top once I figure out how to fix this.